File: scripts/observer_react.ts
import { readFileSync, writeFileSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { evaluateStateFile } from "./equal_rank_nudge_bot";
import { buildActionPlanFromPayload, parseNumber } from "./_observer_tooling";
import { scan } from "./observer_scan";

type ScanTop = {
  state_file: string;
  reason: string;
  message: string | null;
};

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(__dirname, "..");

function nowIso(): string {
  return new Date().toISOString();
}

function readState(stateFile: string): Record<string, unknown> {
  return JSON.parse(readFileSync(stateFile, "utf8")) as Record<string, unknown>;
}

function buildResponse(
  payload: Record<string, unknown>,
  stateFile: string,
  top: ScanTop,
  shouldAct: boolean,
) {
  const actionPlan = buildActionPlanFromPayload(payload, ROOT);
  return {
    should_act: shouldAct,
    job_id: payload.job_id,
    reason: top.reason,
    message: shouldAct ? top.message : null,
    state_file: path.resolve(stateFile),
    spawn_ready_path: payload.spawn_ready_path,
    spawn_ready_json_path: payload.spawn_ready_json_path,
    action_plan: actionPlan,
  };
}

function promote(stateFile: string, top: ScanTop, apply: boolean) {
  const payload = readState(stateFile);
  payload.status = "auto_execute";
  payload.observer = (payload.observer as Record<string, unknown>) ?? {};
  (payload.observer as Record<string, unknown>).last_auto_execute_at = nowIso();
  (payload.observer as Record<string, unknown>).last_auto_execute_reason = top.reason;
  payload.nudge = (payload.nudge as Record<string, unknown>) ?? {};
  (payload.nudge as Record<string, unknown>).proposal_pending = false;

  if (apply) {
    writeFileSync(stateFile, `${JSON.stringify(payload, null, 2)}\n`, "utf8");
  }

  return buildResponse(payload, stateFile, top, true);
}

function reactOne(
  stateFile: string,
  minutesSinceReply: number,
  apply: boolean,
  seed?: number,
) {
  const top = evaluateStateFile(stateFile, minutesSinceReply, seed, apply);
  const payload = readState(stateFile);
  if (!top.should_nudge) {
    return buildResponse(payload, stateFile, top as ScanTop, false);
  }
  return promote(stateFile, top as ScanTop, apply);
}

function react(apply: boolean, seed?: number) {
  const result = scan(apply, seed);
  const top = result.top;
  if (!top) {
    return {
      should_act: false,
      reason: "eligible 없음",
      message: null,
      action_plan: null,
    };
  }
  return promote(String(top.state_file), top as ScanTop, apply);
}

function parseArgs(argv: string[]): {
  apply: boolean;
  seed?: number;
  stateFile?: string;
  minutesSinceReply: number;
} {
  const toFind = (key: string): string | undefined => {
    const idx = argv.indexOf(`--${key}`);
    if (idx === -1 || idx + 1 >= argv.length) return undefined;
    return argv[idx + 1];
  };
  const has = (key: string): boolean => argv.includes(`--${key}`);
  const minutesSinceReply = toFind("minutes-since-reply") ? Number(toFind("minutes-since-reply")) : 10;
  return {
    apply: has("apply"),
    seed: toFind("seed") ? Number(toFind("seed")) : undefined,
    stateFile: toFind("state-file"),
    minutesSinceReply: parseNumber(minutesSinceReply, 10),
  };
}

function main(): void {
  const args = parseArgs(process.argv.slice(2));
  const result = args.stateFile
    ? reactOne(args.stateFile, args.minutesSinceReply, args.apply, args.seed)
    : react(args.apply, args.seed);
  process.stdout.write(`${JSON.stringify(result, null, 2)}\n`);
}

if (import.meta.main) {
  main();
}

File: scripts/observer_react.py
#!/usr/bin/env python3
import argparse
import json
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Dict, Optional

from observer_scan import scan
from equal_rank_nudge_bot import evaluate_state_file


def now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


def promote(state_file: Path, top: Dict[str, Any], apply: bool) -> Dict[str, Any]:
    payload = json.loads(state_file.read_text(encoding="utf-8"))
    payload["status"] = "auto_execute"
    payload.setdefault("observer", {})
    payload["observer"]["last_auto_execute_at"] = now_iso()
    payload["observer"]["last_auto_execute_reason"] = top["reason"]
    payload.setdefault("nudge", {})
    payload["nudge"]["proposal_pending"] = False

    if apply:
        state_file.write_text(json.dumps(payload, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")

    return {
        "should_act": True,
        "job_id": payload.get("job_id"),
        "reason": top["reason"],
        "message": top["message"],
        "state_file": str(state_file),
        "spawn_ready_path": payload.get("spawn_ready_path"),
    }


def react(apply: bool, seed: Optional[int] = None) -> Dict[str, Any]:
    result = scan(apply=apply, seed=seed)
    top = result.get("top")
    if not top:
        return {
            "should_act": False,
            "reason": "eligible 없음",
            "message": None,
        }

    return promote(Path(top["state_file"]), top, apply)


def react_one(state_file: Path, minutes_since_reply: int, apply: bool, seed: Optional[int] = None) -> Dict[str, Any]:
    top = evaluate_state_file(
        state_file=state_file,
        minutes_since_reply=minutes_since_reply,
        seed=seed,
        apply=apply,
    )
    if not top.get("should_nudge"):
        return {
            "should_act": False,
            "reason": top["reason"],
            "message": None,
            "state_file": str(state_file),
        }
    return promote(state_file, top, apply)


def main() -> None:
    p = argparse.ArgumentParser(description="오 분 관찰자 반응기")
    p.add_argument("--apply", action="store_true")
    p.add_argument("--seed", type=int)
    p.add_argument("--state-file")
    p.add_argument("--minutes-since-reply", type=int, default=10)
    args = p.parse_args()

    if args.state_file:
        result = react_one(Path(args.state_file), args.minutes_since_reply, args.apply, args.seed)
    else:
        result = react(apply=args.apply, seed=args.seed)
    print(json.dumps(result, ensure_ascii=False))


if __name__ == "__main__":
    main()
